File: web/ViewModels/SearchViewModel.cs
using web.Models;
using web.Services;

namespace web.ViewModels
{
    public enum SearchState { Idle, Loading, Success, Empty, Error }
    public class SearchViewModel
    {
        private readonly ApiService _apiService;

        public SearchState State { get; private set; } = SearchState.Idle;
        public List<ArticleModel> Results { get; private set; } = new();
        public string ErrorMessage { get; private set; } = string.Empty;
        public string LastQuery { get; private set; } = string.Empty;
        public bool HasMore { get; private set; } = true;

        private int _currentPage = 1;
        private CancellationTokenSource? _debounceToken;

        public event Action? OnChange;

        public SearchViewModel(ApiService apiService)
        {
            _apiService = apiService;
        }

        public async Task SearchAsync(string query)
        {
            var trimmed = query.Trim();

            if (string.IsNullOrEmpty(trimmed))
            {
                ClearSearch();
                return;
            }
            if (trimmed == LastQuery && State == SearchState.Success) return;

            _debounceToken?.Cancel();
            _debounceToken = new CancellationTokenSource();
            var token = _debounceToken.Token;

            try
            {
                await Task.Delay(500, token);
            }
            catch (TaskCanceledException)
            {
                return;
            }

            LastQuery = trimmed;
            _currentPage = 1;
            HasMore = true;
            Results = new List<ArticleModel>();
            SetState(SearchState.Loading);

            try
            {
                var results = await _apiService.SearchArticlesAsync(trimmed, _currentPage);
                Results = results;
                _currentPage++;
                SetState(Results.Count == 0 ? SearchState.Empty : SearchState.Success);

            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(SearchState.Error);
            }
        }
        public async Task LoadMoreAsync()
        {
            if (!HasMore || State == SearchState.Loading) return;
            SetState(SearchState.Loading);
            try
            {
                var newResults = await _apiService.SearchArticlesAsync(LastQuery, _currentPage);
                Results = new List<ArticleModel>(Results);
                Results.AddRange(newResults);
                HasMore = newResults.Count > 0;
                _currentPage++;

                SetState(SearchState.Success);
            } catch(Exception ex)
            {
                ErrorMessage = ex.Message;
                SetState(SearchState.Error);
            }
        }

        public void ClearSearch()
        {
            _debounceToken?.Cancel();
            Results = new List<ArticleModel>();
            LastQuery = string.Empty;
            _currentPage = 1;
            HasMore = true;
            SetState(SearchState.Idle);
        }
        private void SetState(SearchState state)
        {
            State = state;
            NotifyStateChanged();
        }
        private void NotifyStateChanged() => OnChange?.Invoke(); 
    }
}

File: web/js/viewmodels/SearchViewModel.js
import { searchArticles } from '../services/apiService';
import { Article } from '../models/Article';
import { debounce } from '../utils/helpers';

export class SearchViewModel {
    #results = [];
    #state = 'idle';
    #lastQuery = '';
    #currentPage = 1;
    #hasMore = true;
    #errorMessage = '';
    #listeners = new Set();

    get results() { return this.#results; }
    get state() { return this.#state; }
    get lastQuery() { return this.#lastQuery; }
    get hasMore() { return this.#hasMore; }
    get errorMessage() { return this.#errorMessage; }

    subscribe(fn) {
        this.#listeners.add(fn);
        return () => this.#listeners.delete(fn);
    }

    #notify() {
        this.#listeners.forEach((fn) => fn(this));
    }

    #setState(state) {
        this.#state = state;
        this.#notify();
    }

    search = debounce(async(query) => {
        const trimmmed = query.trim();
        if(!trimmmed) {
            this.clearSearch();
            return;
        }

        if(trimmmed === this.#lastQuery && this.#state === 'success') return;

        this.#lastQuery = trimmmed;
        this.#currentPage = 1;
        this.#hasMore = true;
        this.#results = [];
        this.#setState('loading');

        try {
            const response = await searchArticles(trimmmed, this.#currentPage);
            this.#results = Article.fromList(response.items);
            this.#hasMore = response.hasMore ?? false;
            this.#currentPage++;
            this.#setState(this.#results.length === 0 ? 'empty': 'success');
        } catch(err) {
            this.#errorMessage = err.message;
            this.#setState('error');
        }
    }, 500);

    async loadMore() {
        if(!this.#hasMore || this.#state === 'loading') return;
        this.#setState('loading');

        try {
            const response = await searchArticles(this.#lastQuery, this.#currentPage);
            const newResults = Article.fromList(response.items);
            this.#results = [...this.#results, ...newResults];
            this.#hasMore = newResults.length > 0;
            this.#currentPage++;
            this.#setState('success');
        } catch(err) {
            this.#errorMessage = err.message;
            this.#setState('error');
        }
    }

    clearSearch() {
        this.#results = [];
        this.#lastQuery = '';
        this.#currentPage = 1;
        this.#hasMore = true;
        this.#setState('idle');
    }
}